File: ios/Swift/SoftEtherClient.swift
//
//  SoftEtherClient.swift
//  Integration Layer
//
//  Swift wrapper for SoftEtherZig C FFI
//

import Foundation

/// Connection state
public enum ConnectionState: Int {
    case idle = 0
    case connecting = 1
    case established = 2
    case disconnecting = 3
    case error = 4
}

/// Event level
public enum EventLevel: Int {
    case info = 0
    case warning = 1
    case error = 2
}

/// VPN Configuration
public struct VPNConfig: Codable {
    let serverName: String
    let serverPort: UInt16
    let hubName: String
    let username: String
    let password: String
    let useEncrypt: Bool
    let useCompress: Bool
    let ipVersion: String?
    let maxConnection: Int?
    
    enum CodingKeys: String, CodingKey {
        case serverName = "server_name"
        case serverPort = "server_port"
        case hubName = "hub_name"
        case username
        case password
        case useEncrypt = "use_encrypt"
        case useCompress = "use_compress"
        case ipVersion = "ip_version"
        case maxConnection = "max_connection"
    }
    
    public init(
        serverName: String,
        serverPort: UInt16 = 443,
        hubName: String,
        username: String,
        password: String,
        useEncrypt: Bool = true,
        useCompress: Bool = true,
        ipVersion: String? = "auto",
        maxConnection: Int? = 1
    ) {
        self.serverName = serverName
        self.serverPort = serverPort
        self.hubName = hubName
        self.username = username
        self.password = password
        self.useEncrypt = useEncrypt
        self.useCompress = useCompress
        self.ipVersion = ipVersion
        self.maxConnection = maxConnection
    }
}

/// Network settings from VPN server
public struct NetworkSettings: Codable {
    let assignedIPv4: String?
    let subnetMask: String?
    let gateway: String?
    let dnsServers: [String]?
    
    enum CodingKeys: String, CodingKey {
        case assignedIPv4 = "assigned_ipv4"
        case subnetMask = "subnet_mask"
        case gateway
        case dnsServers = "dns_servers"
    }
}

/// Connection statistics
public struct ConnectionStats {
    let bytesSent: UInt64
    let bytesReceived: UInt64
    let packetsSent: UInt64
    let packetsReceived: UInt64
    let connectedSeconds: UInt64
    let currentRttMs: UInt32
}

/// VPN Error types
public enum VPNError: Error {
    case notInitialized
    case initializationFailed
    case connectionFailed(code: Int)
    case disconnectionFailed
    case invalidConfiguration
    case operationFailed(message: String)
    case notConnected
}

/// Main SoftEther client class
public class SoftEtherClient {
    private var handle: OpaquePointer?
    private var stateCallback: ((ConnectionState) -> Void)?
    private var eventCallback: ((EventLevel, Int, String) -> Void)?
    private var ipPacketCallback: ((Data) -> Void)?
    private var logCallback: ((Date, Int, String, String) -> Void)?
    
    // Context pointers for C callbacks
    private var stateContext: UnsafeMutableRawPointer?
    private var eventContext: UnsafeMutableRawPointer?
    private var ipRxContext: UnsafeMutableRawPointer?
    private var logContext: UnsafeMutableRawPointer?
    
    public init() {
        // Set default log level
        set_log_level(LOG_LEVEL_INFO)
    }
    
    deinit {
        disconnect()
        if let handle = handle {
            softether_client_free(handle)
        }
        cleanupContexts()
    }
    
    /// Create client with configuration
    public func create(config: VPNConfig) throws {
        // Encode config to JSON
        let encoder = JSONEncoder()
        let jsonData = try encoder.encode(config)
        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw VPNError.invalidConfiguration
        }
        
        // Create client
        handle = softether_client_create_for_ios(jsonString)
        guard handle != nil else {
            throw VPNError.initializationFailed
        }
    }
    
    /// Connect to VPN server
    public func connect() throws {
        guard let handle = handle else {
            throw VPNError.notInitialized
        }
        
        let result = softether_client_connect(handle)
        if result != 0 {
            let errorMsg = getLastError()
            throw VPNError.connectionFailed(code: Int(result))
        }
    }
    
    /// Disconnect from VPN server
    public func disconnect() {
        guard let handle = handle else { return }
        _ = softether_client_disconnect(handle)
    }
    
    /// Get current connection state
    public func getState() -> ConnectionState {
        guard let handle = handle else {
            return .idle
        }
        let state = softether_client_get_state(handle)
        return ConnectionState(rawValue: Int(state)) ?? .error
    }
    
    /// Check if connected
    public func isConnected() -> Bool {
        guard let handle = handle else { return false }
        return softether_client_is_connected(handle) != 0
    }
    
    /// Get connection statistics
    public func getStats() throws -> ConnectionStats {
        guard let handle = handle else {
            throw VPNError.notInitialized
        }
        
        var stats = softether_connection_stats_t()
        let result = softether_client_get_stats(handle, &stats)
        
        if result != 0 {
            throw VPNError.operationFailed(message: "Failed to get statistics")
        }
        
        return ConnectionStats(
            bytesSent: stats.bytes_sent,
            bytesReceived: stats.bytes_received,
            packetsSent: stats.packets_sent,
            packetsReceived: stats.packets_received,
            connectedSeconds: stats.connected_seconds,
            currentRttMs: stats.current_rtt_ms
        )
    }
    
    /// Get network settings from server
    public func getNetworkSettings() throws -> NetworkSettings {
        guard let handle = handle else {
            throw VPNError.notInitialized
        }
        
        guard let jsonPtr = softether_client_get_network_settings_json(handle) else {
            throw VPNError.operationFailed(message: "Failed to get network settings")
        }
        defer { softether_string_free(jsonPtr) }
        
        let jsonString = String(cString: jsonPtr)
        let jsonData = jsonString.data(using: .utf8)!
        
        let decoder = JSONDecoder()
        return try decoder.decode(NetworkSettings.self, from: jsonData)
    }
    
    /// Get last error message
    public func getLastError() -> String {
        guard let handle = handle else {
            return "Client not initialized"
        }
        
        guard let errorPtr = softether_client_last_error(handle) else {
            return "No error"
        }
        defer { softether_string_free(errorPtr) }
        
        return String(cString: errorPtr)
    }
    
    /// Set state change callback
    public func setStateCallback(_ callback: @escaping (ConnectionState) -> Void) {
        self.stateCallback = callback
        
        guard let handle = handle else { return }
        
        // Create context
        let context = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())
        stateContext = context
        
        // C callback wrapper
        let cCallback: @convention(c) (Int32, UnsafeMutableRawPointer?) -> Void = { state, userPtr in
            guard let userPtr = userPtr else { return }
            let client = Unmanaged<SoftEtherClient>.fromOpaque(userPtr).takeUnretainedValue()
            
            if let state = ConnectionState(rawValue: Int(state)) {
                DispatchQueue.main.async {
                    client.stateCallback?(state)
                }
            }
        }
        
        softether_client_set_state_callback(handle, cCallback, context)
    }
    
    /// Set event callback
    public func setEventCallback(_ callback: @escaping (EventLevel, Int, String) -> Void) {
        self.eventCallback = callback
        
        guard let handle = handle else { return }
        
        let context = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())
        eventContext = context
        
        let cCallback: @convention(c) (Int32, Int32, UnsafePointer<CChar>?, UnsafeMutableRawPointer?) -> Void = {
            level, code, message, userPtr in
            guard let userPtr = userPtr else { return }
            let client = Unmanaged<SoftEtherClient>.fromOpaque(userPtr).takeUnretainedValue()
            
            let levelEnum = EventLevel(rawValue: Int(level)) ?? .info
            let msg = message != nil ? String(cString: message!) : ""
            
            DispatchQueue.main.async {
                client.eventCallback?(levelEnum, Int(code), msg)
            }
        }
        
        softether_client_set_event_callback(handle, cCallback, context)
    }
    
    /// Set IP packet receive callback (for PacketTunnelProvider)
    public func setIPPacketCallback(_ callback: @escaping (Data) -> Void) {
        self.ipPacketCallback = callback
        
        guard let handle = handle else { return }
        
        let context = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())
        ipRxContext = context
        
        let cCallback: @convention(c) (UnsafePointer<UInt8>?, UInt32, UnsafeMutableRawPointer?) -> Void = {
            dataPtr, len, userPtr in
            guard let userPtr = userPtr, let dataPtr = dataPtr else { return }
            let client = Unmanaged<SoftEtherClient>.fromOpaque(userPtr).takeUnretainedValue()
            
            let data = Data(bytes: dataPtr, count: Int(len))
            
            DispatchQueue.main.async {
                client.ipPacketCallback?(data)
            }
        }
        
        softether_client_set_ip_rx_callback(handle, cCallback, context)
    }
    
    /// Send IP packet (for PacketTunnelProvider)
    public func sendIPPacket(_ packet: Data) -> Bool {
        guard let handle = handle else { return false }
        
        let result = packet.withUnsafeBytes { bufferPtr in
            guard let baseAddress = bufferPtr.baseAddress else { return Int32(-1) }
            return softether_client_send_ip_packet(
                handle,
                baseAddress.assumingMemoryBound(to: UInt8.self),
                UInt32(packet.count)
            )
        }
        
        return result > 0
    }
    
    /// Set reconnection enabled
    public func setReconnectEnabled(_ enabled: Bool) {
        guard let handle = handle else { return }
        _ = softether_client_set_reconnect_enabled(handle, enabled ? 1 : 0)
    }
    
    /// Set reconnection parameters
    public func setReconnectParams(maxAttempts: UInt32, initialDelay: UInt32, maxDelay: UInt32) {
        guard let handle = handle else { return }
        _ = softether_client_set_reconnect_params(handle, maxAttempts, initialDelay, maxDelay)
    }
    
    /// Set log level
    public func setLogLevel(_ level: LogLevel) {
        set_log_level(level)
    }
    
    /// Get version string
    public static func getVersion() -> String {
        guard let versionPtr = softether_client_version() else {
            return "Unknown"
        }
        defer { softether_string_free(versionPtr) }
        return String(cString: versionPtr)
    }
    
    // MARK: - Private helpers
    
    private func cleanupContexts() {
        if let ctx = stateContext {
            Unmanaged<SoftEtherClient>.fromOpaque(ctx).release()
        }
        if let ctx = eventContext {
            Unmanaged<SoftEtherClient>.fromOpaque(ctx).release()
        }
        if let ctx = ipRxContext {
            Unmanaged<SoftEtherClient>.fromOpaque(ctx).release()
        }
        if let ctx = logContext {
            Unmanaged<SoftEtherClient>.fromOpaque(ctx).release()
        }
    }
}

// MARK: - Utility Extensions

extension SoftEtherClient {
    /// Test connectivity to a server
    public static func testConnectivity(
        serverName: String,
        serverPort: UInt16 = 443,
        timeoutMs: UInt32 = 5000
    ) -> Bool {
        let result = softether_client_test_connectivity(serverName, serverPort, timeoutMs)
        return result == 0
    }
    
    /// Resolve DNS
    public static func resolveDNS(hostname: String, dnsServer: String? = nil) throws -> [String] {
        let dns = dnsServer ?? "8.8.8.8"
        
        guard let jsonPtr = softether_dns_resolve(hostname, dns) else {
            throw VPNError.operationFailed(message: "DNS resolution failed")
        }
        defer { softether_string_free(jsonPtr) }
        
        let jsonString = String(cString: jsonPtr)
        let jsonData = jsonString.data(using: .utf8)!
        
        struct DNSResult: Codable {
            let addresses: [String]
            let cname: String?
        }
        
        let decoder = JSONDecoder()
        let result = try decoder.decode(DNSResult.self, from: jsonData)
        return result.addresses
    }
}

File: android/kotlin/SoftEtherClient.kt
package com.softether.vpn

import android.util.Log
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Connection state enum matching C API
 */
enum class ConnectionState(val value: Int) {
    IDLE(0),
    CONNECTING(1),
    ESTABLISHED(2),
    DISCONNECTING(3),
    ERROR(4);
    
    companion object {
        fun fromInt(value: Int) = values().firstOrNull { it.value == value } ?: ERROR
    }
}

/**
 * Event level enum
 */
enum class EventLevel(val value: Int) {
    INFO(0),
    WARNING(1),
    ERROR(2);
    
    companion object {
        fun fromInt(value: Int) = values().firstOrNull { it.value == value } ?: INFO
    }
}

/**
 * VPN Configuration data class
 */
data class VPNConfig(
    val serverName: String,
    val serverPort: Int = 443,
    val hubName: String,
    val username: String,
    val password: String,
    val useEncrypt: Boolean = true,
    val useCompress: Boolean = true,
    val ipVersion: String = "auto",
    val maxConnection: Int = 1
) {
    fun toJson(): String {
        return JSONObject().apply {
            put("server_name", serverName)
            put("server_port", serverPort)
            put("hub_name", hubName)
            put("username", username)
            put("password", password)
            put("use_encrypt", useEncrypt)
            put("use_compress", useCompress)
            put("ip_version", ipVersion)
            put("max_connection", maxConnection)
        }.toString()
    }
}

/**
 * Network settings from VPN server
 */
data class NetworkSettings(
    val assignedIPv4: String?,
    val subnetMask: String?,
    val gateway: String?,
    val dnsServers: List<String>?
)

/**
 * Connection statistics
 */
data class ConnectionStats(
    val bytesSent: Long,
    val bytesReceived: Long,
    val packetsSent: Long,
    val packetsReceived: Long,
    val connectedSeconds: Long,
    val currentRttMs: Int
)

/**
 * VPN Error exceptions
 */
sealed class VPNException(message: String) : Exception(message) {
    class NotInitialized : VPNException("Client not initialized")
    class InitializationFailed : VPNException("Failed to initialize client")
    class ConnectionFailed(val code: Int) : VPNException("Connection failed with code $code")
    class DisconnectionFailed : VPNException("Disconnection failed")
    class InvalidConfiguration : VPNException("Invalid configuration")
    class OperationFailed(message: String) : VPNException(message)
    class NotConnected : VPNException("Not connected to VPN")
}

/**
 * Main SoftEther Client class for Android
 */
class SoftEtherClient {
    
    private var handle: Long = 0
    private val initialized = AtomicBoolean(false)
    
    // Callbacks
    private var stateCallback: ((ConnectionState) -> Unit)? = null
    private var eventCallback: ((EventLevel, Int, String) -> Unit)? = null
    private var logCallback: ((Long, Int, String, String) -> Unit)? = null
    
    companion object {
        private const val TAG = "SoftEtherClient"
        
        init {
            try {
                System.loadLibrary("softether")
                Log.i(TAG, "SoftEther native library loaded")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load SoftEther library", e)
                throw e
            }
        }
        
        /**
         * Get library version
         */
        @JvmStatic
        external fun getVersion(): String
        
        /**
         * Test connectivity to a server
         */
        @JvmStatic
        external fun testConnectivity(
            serverName: String,
            serverPort: Int,
            timeoutMs: Int
        ): Boolean
        
        /**
         * Resolve DNS
         */
        @JvmStatic
        external fun resolveDNS(hostname: String, dnsServer: String): List<String>?
    }
    
    // Native methods
    private external fun nativeCreate(jsonConfig: String): Long
    private external fun nativeFree(handle: Long)
    private external fun nativeConnect(handle: Long): Int
    private external fun nativeDisconnect(handle: Long): Int
    private external fun nativeGetState(handle: Long): Int
    private external fun nativeIsConnected(handle: Long): Boolean
    private external fun nativeGetStats(handle: Long): LongArray?
    private external fun nativeGetNetworkSettings(handle: Long): String?
    private external fun nativeGetLastError(handle: Long): String
    private external fun nativeSetStateCallback(handle: Long)
    private external fun nativeSetEventCallback(handle: Long)
    private external fun nativeSetReconnectEnabled(handle: Long, enabled: Boolean)
    private external fun nativeSetReconnectParams(
        handle: Long,
        maxAttempts: Int,
        initialDelay: Int,
        maxDelay: Int
    )
    private external fun nativeSetLogLevel(level: Int)
    private external fun nativeGetTunFd(handle: Long): Int
    private external fun nativeSetTunFd(handle: Long, fd: Int): Int
    
    /**
     * Create client with configuration
     */
    fun create(config: VPNConfig) {
        if (initialized.get()) {
            throw VPNException.OperationFailed("Client already initialized")
        }
        
        val jsonConfig = config.toJson()
        handle = nativeCreate(jsonConfig)
        
        if (handle == 0L) {
            throw VPNException.InitializationFailed()
        }
        
        initialized.set(true)
        Log.i(TAG, "Client created successfully")
    }
    
    /**
     * Clean up resources
     */
    fun destroy() {
        if (!initialized.get()) return
        
        disconnect()
        
        if (handle != 0L) {
            nativeFree(handle)
            handle = 0
        }
        
        initialized.set(false)
        Log.i(TAG, "Client destroyed")
    }
    
    /**
     * Connect to VPN server
     */
    fun connect() {
        if (!initialized.get()) {
            throw VPNException.NotInitialized()
        }
        
        val result = nativeConnect(handle)
        if (result != 0) {
            val error = getLastError()
            Log.e(TAG, "Connection failed: $error")
            throw VPNException.ConnectionFailed(result)
        }
        
        Log.i(TAG, "Connection initiated")
    }
    
    /**
     * Disconnect from VPN server
     */
    fun disconnect() {
        if (!initialized.get()) return
        
        val result = nativeDisconnect(handle)
        if (result != 0) {
            Log.w(TAG, "Disconnection returned code $result")
        }
        
        Log.i(TAG, "Disconnected")
    }
    
    /**
     * Get current connection state
     */
    fun getState(): ConnectionState {
        if (!initialized.get()) {
            return ConnectionState.IDLE
        }
        
        val state = nativeGetState(handle)
        return ConnectionState.fromInt(state)
    }
    
    /**
     * Check if connected
     */
    fun isConnected(): Boolean {
        if (!initialized.get()) return false
        return nativeIsConnected(handle)
    }
    
    /**
     * Get connection statistics
     */
    fun getStats(): ConnectionStats {
        if (!initialized.get()) {
            throw VPNException.NotInitialized()
        }
        
        val stats = nativeGetStats(handle)
            ?: throw VPNException.OperationFailed("Failed to get statistics")
        
        return ConnectionStats(
            bytesSent = stats[0],
            bytesReceived = stats[1],
            packetsSent = stats[2],
            packetsReceived = stats[3],
            connectedSeconds = stats[4],
            currentRttMs = stats[5].toInt()
        )
    }
    
    /**
     * Get network settings from server
     */
    fun getNetworkSettings(): NetworkSettings {
        if (!initialized.get()) {
            throw VPNException.NotInitialized()
        }
        
        val jsonStr = nativeGetNetworkSettings(handle)
            ?: throw VPNException.OperationFailed("Failed to get network settings")
        
        val json = JSONObject(jsonStr)
        return NetworkSettings(
            assignedIPv4 = json.optString("assigned_ipv4", null),
            subnetMask = json.optString("subnet_mask", null),
            gateway = json.optString("gateway", null),
            dnsServers = json.optJSONArray("dns_servers")?.let { array ->
                (0 until array.length()).map { array.getString(it) }
            }
        )
    }
    
    /**
     * Get last error message
     */
    fun getLastError(): String {
        if (!initialized.get()) {
            return "Client not initialized"
        }
        return nativeGetLastError(handle)
    }
    
    /**
     * Set state change callback
     */
    fun setStateCallback(callback: (ConnectionState) -> Unit) {
        stateCallback = callback
        if (initialized.get()) {
            nativeSetStateCallback(handle)
        }
    }
    
    /**
     * Set event callback
     */
    fun setEventCallback(callback: (EventLevel, Int, String) -> Unit) {
        eventCallback = callback
        if (initialized.get()) {
            nativeSetEventCallback(handle)
        }
    }
    
    /**
     * Set reconnection enabled
     */
    fun setReconnectEnabled(enabled: Boolean) {
        if (initialized.get()) {
            nativeSetReconnectEnabled(handle, enabled)
        }
    }
    
    /**
     * Set reconnection parameters
     */
    fun setReconnectParams(maxAttempts: Int, initialDelay: Int, maxDelay: Int) {
        if (initialized.get()) {
            nativeSetReconnectParams(handle, maxAttempts, initialDelay, maxDelay)
        }
    }
    
    /**
     * Set log level
     */
    fun setLogLevel(level: Int) {
        nativeSetLogLevel(level)
    }
    
    /**
     * Get tunnel file descriptor (for VpnService)
     */
    fun getTunFd(): Int {
        if (!initialized.get()) {
            throw VPNException.NotInitialized()
        }
        return nativeGetTunFd(handle)
    }
    
    /**
     * Set tunnel file descriptor (for VpnService)
     */
    fun setTunFd(fd: Int) {
        if (!initialized.get()) {
            throw VPNException.NotInitialized()
        }
        val result = nativeSetTunFd(handle, fd)
        if (result != 0) {
            throw VPNException.OperationFailed("Failed to set tunnel FD")
        }
    }
    
    // Called from JNI
    @Suppress("unused")
    private fun onStateChanged(state: Int) {
        val connState = ConnectionState.fromInt(state)
        Log.d(TAG, "State changed: $connState")
        stateCallback?.invoke(connState)
    }
    
    // Called from JNI
    @Suppress("unused")
    private fun onEvent(level: Int, code: Int, message: String) {
        val eventLevel = EventLevel.fromInt(level)
        Log.d(TAG, "Event [$eventLevel] code=$code: $message")
        eventCallback?.invoke(eventLevel, code, message)
    }
    
    // Called from JNI
    @Suppress("unused")
    private fun onLog(timestamp: Long, level: Int, source: String, message: String) {
        logCallback?.invoke(timestamp, level, source, message)
    }
}
